contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,781 @@
+-  Allow SchedulingPolicy to be used by pods using RoleBindings or ClusterRoleBindings
+
+# Detailed Design
+
+SchedulingPolicy resource specs are composed of several main attributes:
+-  **Required** scheduling components. These list the manadory NodeSelectors, Scheduler-Name, Anti-Affinity rules, Priority class, and Tolerations and optionally valid values that have to be provided to allow scheduling.
+-  **Allowed** scheduling components. These list the optional components that can be specified in pods definition.
+-  **Default** scheduling components. These list default values to set unless specified in pods definition.
+
+SchedulingPolicy resources are supposed to apply in a deny-all-except approach.
+
+An AdmissionController must be added to the mutating phase to
+-  add default values if unspecified for NodeSelectors, Scheduler-Name, Anti-Affinity rules, Priority class, and Tolerations,
+-  reject pod scheduling if the serviceaccount running the pod is not allowed to specify requested NodeSelectors, Scheduler-Name, Anti-Affinity rules, Priority class, and Tolerations.
+
+All usable scheduling policies (allowed by RBAC) are merged before evaluating if scheduling constraints defined in pods are allowed.
+
+## SchedulingPolicy
+
+Proposed API group: `extensions/v1alpha1`
+
+SchedulingPolicy is a cluster-scoped resource (not namespaced).
+
+### SchedulingPolicy content
+
+SchedulingPolicy spec is composed of optional fields that allow scheduling rules. If a field is absent from a SchedulingPolicy, this schedpol won't allow any item from the missing field.
+
+```yaml
+apiVersion: extensions/valpha1
+kind: SchedulingPolicy
+metadata:
+  name: my-schedpol
+spec:
+  required:
+    schedulerNames:      # Describes schedulers names that are required
+    priorityClasseNames: # Describes priority class names that are required
+    nodeSelectors:       # Describes node selectors that must be used
+    affinities:          # Describes affinities that must be used
+  allowed:
+    schedulerNames:      # Describes schedulers names that are allowed
+    priorityClasseNames: # Describes priority class names that are allowed
+    nodeSelectors:       # Describes node selectors that can be used
+    tolerations:         # Describes tolerations that can be used
+    affinities:          # Describes affinities that can be used
+  default:
+    schedulerName:       # Describes default scheduler name
+    priorityClasseName:  # Describes default priority class name
+    nodeSelector:        # Describes default node selector
+    tolerations:         # Describes default tolerations
+    affinity:            # Describes default affinity
+```
+
+### required
+Elements here are required, pods won't schedule if they aren't present. Also note that if something is required it is also allowed.
+
+### allowed
+Elements here are allowed, the policy allows the presence of these elements. From Pod's perspective, a pod can use one or N of the allowed items.
```

Comment:
We should think about backward compatibility, rollout, and UX as well. I think we should allow all by default as long as there is no "allow" rule. As soon as there is one allow rule, it means only allow those properties specified in the rule.

Reply:
>  I think we should allow all by default as long as there is no "allow" rule. As soon as there is one allow rule, it means only allow those properties specified in the rule.

What should the posture of the policy be toward new future capabilities? If I make a policy that locks down scheduling in a specific way, and a new field in a future version lets pods target scheduling in a different way, wouldn't you expect the old policy to keep them as limited as they were before the new field existed? If so, that's a default deny approach. That approach makes rolling out new capabilities much easier/safer. 

I agree that default deny makes rolling out on top of existing clusters with existing features harder, but without that approach, every upgrade makes admins have to ask "what new features/holes got introduced I have to update my policies to cover/block?"

Reply:
I agree with @liggitt as users do not expect re-think their policies every upgrade.

Reply:
> What should the posture of the policy be toward new future capabilities? If I make a policy that locks down scheduling in a specific way, and a new field in a future version lets pods target scheduling in a different way, wouldn't you expect the old policy to keep them as limited as they were before the new field existed? If so, that's a default deny approach. That approach makes rolling out new capabilities much easier/safer.

As I explained in my previous comment, once you have one "allow" rule, it means that nothing else is allowed, but when you don't have any "allow" rule, it means everything is allowed. In other words, the policy is "default allow" when there is no "allow" rule and it changes to "default deny" once one or more "allow" rule is added. This would solve the problem of adding a new scheduling feature and makes rolling out pod scheduling policy easy.

Reply:
I would like the points discussed here to be explicitly reflected in the design doc.